integration-tests/byoid/integration_test.go

```diff
@@ -231,3 +240,84 @@ func TestURLBasedCredentials(t *testing.T) {
 		},
 	})
 }
+
+// Tests to make sure AWS based external credentials work properly.
+func TestAWSBasedCredentials(t *testing.T) {
+	data := neturl.Values{}
+	data.Set("audience", clientID)
+	data.Set("includeEmail", "true")
+
+	client, err := google.DefaultClient(context.Background(), "https://www.googleapis.com/auth/cloud-platform")
+	resp, err := client.PostForm(fmt.Sprintf("https://iamcredentials.googleapis.com/v1/%s:generateIdToken", clientID), data)
+	if resp.StatusCode != 200 {
+		t.Errorf("byoid: failed to get Google ID token for AWS test: %v", err)
+	}
+
+	var res map[string]interface{}
+
+	json.NewDecoder(resp.Body).Decode(&res)
+
+	token, ok := res["token"]
+	if !ok {
+		t.Errorf("byoid: Didn't receieve an ID token back from generateIDToken")
+	}
+
+	data = neturl.Values{}
+	data.Set("Action", "AssumeRoleWithWebIdentity")
+	data.Set("Version", "2011-06-15")
+	data.Set("DurationSeconds", "3600")
+	data.Set("RoleSessionName", os.Getenv("GCLOUD_TESTS_GOLANG_AWS_ROLE_NAME"))
+	data.Set("RoleArn", os.Getenv("GCLOUD_TESTS_GOLANG_AWS_ROLE_ID"))
+	data.Set("WebIdentityToken", token.(string))
+
+	resp, err = http.PostForm("https://sts.amazonaws.com/", data)
+	bodyBytes, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		t.Errorf("byoid: failed to parse response body from AWS: %v", err)
+	}
+
+	bodyString := string(bodyBytes)
+
+	// Using regex instead of loading the full XML response because I only need a few fields.
+	SessTok := regexp.MustCompile("<SessionToken>(.*)</SessionToken>")
```

Comment:
You should not need to load the full payload. Just define a minimal struct with xml struct tags then use the xml package to parse.

Reply:
+
if you wish, I can go learn the XML package and use that instead.  I thought this was avoiding loading the full payload, honesty, but maybe I misinterpreted what that meant.  I thought that I'd have to define an XML struct with tags supporting everything in the response in order to do this, which would be quite a lot.  I tried to model  this off of the Python implementation.

Reply:
I think the test would be cleaner if the xml package was used. The package has some great examples, such as this one: https://pkg.go.dev/encoding/xml#example-Unmarshal. It should have the same feel as the json package. You only need to define the fields that you wish to parse.